### PR TITLE
[FLINK-30998] Add optional exception handler to flink-connector-opensearch

### PR DESCRIPTION
**Pull request by lilyevsky**

### CODE REVIEW

**Comment by reta** [comment]
@lilyevsky the relevant changes looks good, but your branch is out of sync with https://github.com/apache/flink-connector-opensearch, you need to sync + rebase

**Comment by lilyevsky** [comment]
@reta I guess I did something wrong with my repo; I made a minor change in the comments, so I wanted to squish the commits and force push. I did rebase, I thought, against your commit that I saw in the history. I thought then PR would only show my one commit.
It now shows some other commits that, I believe, were already merged into apache:main.
Please let me know if this is a problem. Worst case, I can do it all over again clean.

**Comment by lilyevsky** [comment]
I think I can do it in github: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork

File: flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchSink.java
Comment (reta) [inline flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchSink.java:77]:
```suggestion
        this.failureHandler = checkNotNull(failureHandler);
```

File: flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchSinkBuilder.java
Comment (reta) [inline flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchSinkBuilder.java:75]:
```suggestion
    private FailureHandler failureHandler = DEFAULT_FAILURE_HANDLER; 
```

File: flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchWriter.java
Comment (reta) [inline flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchWriter.java:351]: This `null` check is not needed anymore (the builder would supply the default handler), just `failureHandler.onFailure(chainedFailures);`

**Comment by reta** [comment]
Thanks @lilyevsky just a few minor comments, I think the only part that is missed are tests, and probably you may need to run `mvn spotless:apply` to apply the formatting, thank you!

**Comment by lilyevsky** [comment]
@reta Did all the changes you suggested, also ran the mvn spotless:apply .
Btw., my IntelliJ showed a list of warnings in places not relevant to my changes. Even one error (which is not really an error because it is in the comments), in OpensearchWriter.java:96.
Do you have any suggestions for the tests? As I mentioned before, I actually tested this in real environment; on the other hand, I understand that unit tests are very important to have.

**Comment by reta** [comment]
> @reta Did all the changes you suggested, also ran the mvn spotless:apply . 

👍  thank you!

>  Even one error (which is not really an error because it is in the comments), in OpensearchWriter.java:96.

Hm .. could you share the error please?

> Do you have any suggestions for the tests? 

Yes, sure, I think [OpensearchWriterITCase](https://github.com/apache/flink-connector-opensearch/blob/main/flink-connector-opensearch/src/test/java/org/apache/flink/connector/opensearch/sink/OpensearchWriterITCase.java) could be a good candidate to add a test case with the simulated failure (fe we could pause / shutdown Opensearch container, just an idea). It may be a bit difficult, please let me know if help needed, thank you.

**Comment by lilyevsky** [comment]
The error is about this line:

 * @param bulkProcessorBuilderFactory configuring the {@link BulkProcessor}'s builder

Apparently that parameter is no longer present. 
Also 10 warnings, like "Return value ... is never used" or "Exception ... is never thrown".
One warning is actually about failureHandler, missing description.

I actually can fix all of that if it is OK, then you can review.

Also will look into the tests tomorrow.

**Comment by reta** [comment]
> I actually can fix all of that if it is OK, then you can review.

Thank you, cleanup would be much appreciated, I would suggest let's finish up with tests first so to have the scope finalized, and we could fix javadocs after. Thanks!

**Comment by lilyevsky** [comment]
Sure. 
I will fix that bulkProcessorBuilderFactory  thing, everything else is less important at the moment. I got that list of warnings while committing. When I just run 'analyze', I get even more stuff.

**Comment by lilyevsky** [comment]
@reta I fixed some warnings, pushed as a separate commit, so you can review.
Now looking at the tests. We need to somehow simulate the Opensearch server that can get the request and then send back the response. Is there currently such thing in the test suite? Is it what you do in flink-connector-opensearch-e2e-tests?
In any case, I think you would know better than I how to do it.

**Comment by reta** [comment]
> @reta I fixed some warnings, pushed as a separate commit, so you can review. Now looking at the tests. We need to somehow simulate the Opensearch server that can get the request and then send back the response. 

I shared the hints with you here https://github.com/apache/flink-connector-opensearch/pull/11#issuecomment-1432308014

Here is sample test case:

```
    @Test
    void testWriteErrorOnUpdate() throws Exception {
        final String index = "test-bulk-flush-with-error";
        final int flushAfterNActions = 1;
        final BulkProcessorConfig bulkProcessorConfig =
                new BulkProcessorConfig(flushAfterNActions, -1, -1, FlushBackoffType.NONE, 0, 0);

        try (final OpensearchWriter<Tuple2<Integer, String>> writer =
                createWriter(index, true, bulkProcessorConfig)) {
            // Trigger an error by updating non-existing document
            writer.write(Tuple2.of(1, "u" + buildMessage(1)), null);
            context.assertThatIdsAreNotWritten(index, 1);
        }
    }
```

You need to attach the failure handler and assert afterwards it has been called.

**Comment by lilyevsky** [comment]
@reta Trying to add the test. For some reason the tests from OpensearchWriterITCase class are not executed at all.
I run the command:

mvn clean test

in flink-connector-opensearch directory. I see that some tests from org.apache.flink.connector.opensearch.sink.OpensearchSinkBuilderTest are executed, but not from other 4 files in the same directory.
What could be wrong?

[maventest.txt](https://github.com/apache/flink-connector-opensearch/files/10760344/maventest.txt)

**Comment by lilyevsky** [comment]
Thanks @reta . One more step. It looks like I need to have a "valid Docker environment". I install the docker things:

sudo apt-get install docker-ce docker-ce-cli containerd.io docker-buildx-plugin docker-compose-plugin

What else do I need to do?

**Comment by lilyevsky** [comment]
@reta I installed docker, the daemon is running, the hello world test worked.
However, the 'mvn verify' still fails with "OpensearchSinkITCase » IllegalState Could not find a valid Docker environment...." ....
How does maven know where to look for Docker environment? Maybe I am some env variable?

**Comment by lilyevsky** [comment]
@reta Added the test. I think now it is good.
I also tested the test itself, by simulating the "bad" handler, making the test fail, just to make sure it checks it properly.

File: flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchWriter.java
Comment (reta) [inline flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchWriter.java:110]: Please update javadocs with `failureHandler` arg

File: flink-connector-opensearch/src/test/java/org/apache/flink/connector/opensearch/sink/OpensearchWriterITCase.java
Comment (reta) [inline flink-connector-opensearch/src/test/java/org/apache/flink/connector/opensearch/sink/OpensearchWriterITCase.java:254]: Super nit, just `@Override`

File: flink-connector-opensearch/src/test/java/org/apache/flink/connector/opensearch/sink/OpensearchWriterITCase.java
Comment (reta) [inline flink-connector-opensearch/src/test/java/org/apache/flink/connector/opensearch/sink/OpensearchWriterITCase.java:243]: Probably `private static` is better

File: flink-connector-opensearch/src/test/java/org/apache/flink/connector/opensearch/sink/OpensearchWriterITCase.java
Comment (reta) [inline flink-connector-opensearch/src/test/java/org/apache/flink/connector/opensearch/sink/OpensearchWriterITCase.java:273]: :+1:

File: flink-connector-opensearch/src/test/java/org/apache/flink/connector/opensearch/sink/OpensearchWriterITCase.java
Comment (reta) [inline flink-connector-opensearch/src/test/java/org/apache/flink/connector/opensearch/sink/OpensearchWriterITCase.java:178]: You probably don't need this change since `createWriter` has an overloaded version without failure handler

File: flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/FailureHandler.java
Comment (reta) [inline flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/FailureHandler.java:26]: `@Public` -> `@PublicEvolving` would be probably better (to leave some space to evolve the API)

**Comment by reta** [comment]
@lilyevsky a few minor comments, LGTM otherwise!

File: flink-connector-opensearch/src/test/java/org/apache/flink/connector/opensearch/sink/OpensearchWriterITCase.java
Comment (lilyevsky) [reply flink-connector-opensearch/src/test/java/org/apache/flink/connector/opensearch/sink/OpensearchWriterITCase.java:178]: Agreed, it will be better not to touch that line. I did it because it needed to pass the metricGroup.
So I am fixing the situation by adding yet another constructor, with metricGroup but without failureHandler.

Comment (reta) [reply flink-connector-opensearch/src/test/java/org/apache/flink/connector/opensearch/sink/OpensearchWriterITCase.java:178]: No need for constructor, `createWriter` is a method in this class

**Comment by lilyevsky** [comment]
@reta I committed the changes you suggested.

File: flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchWriter.java
Comment (reta) [inline flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchWriter.java:135]: Please revert these changes, they violate the `SinkWriter` interface 

[1] https://github.com/apache/flink/blob/master/flink-core/src/main/java/org/apache/flink/api/connector/sink2/SinkWriter.java#L41

File: flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchSinkBuilder.java
Comment (reta) [inline flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchSinkBuilder.java:153]: I don't understand why these suppression are needed?

Comment (lilyevsky) [reply flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchSinkBuilder.java:153]: I put it because compiler gave me a warning about it. I can remove that suppression if you prefer.

Comment (reta) [reply flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchSinkBuilder.java:153]: Please remove, thank you

Comment (lilyevsky) [reply flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchSinkBuilder.java]: Done

Comment (lilyevsky) [reply flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchSinkBuilder.java]: Hi @reta , please let me know what are the plans for merging.

File: flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchWriter.java
Comment (reta) [reply flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchWriter.java:135]: @lilyevsky could you please address this and [this](https://github.com/apache/flink-connector-opensearch/pull/11/files#r1110066551) comment? thank you

Comment (lilyevsky) [reply flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchWriter.java:135]: @reta No problem, could you please clarify: you want me to put back the IOException to both write and flush methods, correct? Also, I am not sure what you mean by addressing  [this](https://github.com/apache/flink-connector-opensearch/pull/11/files#r1110066551) . Is it about removing the "@SuppressWarnings("All")"?
Please confirm.

Comment (reta) [reply flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchWriter.java:135]: @lilyevsky correct, the `throws IOException, InterruptedException` is the right part of the signature: the first comment is about `write`, the second is about `flush`, thank you

Comment (lilyevsky) [reply flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchWriter.java]: @reta Done. Also removed the line with warnings suppressions that I added at some point.

**Review by reta — APPROVED** [review]

**Comment by reta** [comment]
@MartijnVisser @zentol could you please take a look guys? thank you!

File: flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchSinkBuilder.java
Comment (reta) [reply flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchSinkBuilder.java]: > Hi @reta , please let me know what are the plans for merging.

Thanks @lilyevsky , LGTM, asked Flink committers to take I look (I do not have commit rights)

Comment (lilyevsky) [reply flink-connector-opensearch/src/main/java/org/apache/flink/connector/opensearch/sink/OpensearchSinkBuilder.java]: Thanks @reta

**Comment by lilyevsky** [comment]
I see since I finalized this request there were some other changes in main branch, and so there are conflicts.
I am going to rebase my branch and hopefully resolve the conflicts this way.
After I do that, can we merge my request sooner rather than later? Otherwise there might be other conflicts again.

**Comment by lilyevsky** [comment]
I actually resolved the conflicts in the web editor, and updated my branch. 
Can somebody review this?

**Comment by reta** [comment]
@MartijnVisser @snuyanzin please help us here :-)

**Comment by lilyevsky** [comment]
Actually, hold on for now. I was not able to compile it in my work environment, now tried in my home desktop and found some errors. Let me fix them. Will let you know.

**Comment by lilyevsky** [comment]
I made it clean now, you can review and maybe merge already.
I got the latest from apache:main branch and re-applied all my changes on top of it, so you can see now it is just one commit.
I was able to locally build it.

**Review by reta — APPROVED** [review]

**Comment by lilyevsky** [comment]
So you can merge it now?
After you do, I am going to build locally the latest version and do my own tests.

**Pull request merged**
